### ISSUE TITLE
EKS Associated Security Group Deletion Issues

### ISSUE THREAD

**Issue by hangxie** [terraform-providers/terraform-provider-aws#11047]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
```
$ terraform version
Terraform v0.12.16
+ provider.aws v2.40.0
+ provider.external v1.2.0
+ provider.local v1.4.0
+ provider.null v2.1.2
```

### Affected Resource(s)

* aws_eks_cluster

### Terraform Configuration Files

https://gist.github.com/hangxie/092ed9ccb3778d6072437666aaecb0b9

### Debug Output

N/A

### Panic Output

N/A

### Expected Behavior

All resources should be deleted by `terraform destroy`, include the security group created by AWS associated with EKS.

### Actual Behavior

The security group created by AWS associated with EKS is not deleted, which causes `terraform destroy` failed to complete as VPC cannot be deleted due to existing SG.

### Steps to Reproduce

1. `terraform apply`
2. `terraform destroy`

### References

Here are something may be helpful:
1. This does not happen to EKS 1.13 as it does not create cluster security group
2. EKS 1.14 platform version is eks.5, I feel like eks.2 or eks.3 was working as I did not have this problem a week ago. I cannot test EKS 1.14 with other platform version but am willing to do so if there is a way.
3. AWS console and CLI both can delete the EKS cluster cleanly, i.e. the SG was removed, even with the EKS cluster I created with Terraform.

**Comment by bflad** [comment]
Hi @hangxie :wave: Thank you for reporting this and sorry you are running into trouble here.

Given your example configuration:

```hcl
# ... other resources omitted for clarity ...

resource "aws_iam_role" "eks_service_role" {
  name = "eks_service_role"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": [
          "eks.amazonaws.com"
        ]
      },
      "Action": [
        "sts:AssumeRole"
      ]
    }
  ]
}
EOF
}

resource "aws_iam_role_policy_attachment" "eks_service_role_AmazonEKSClusterPolicy" {
  role = aws_iam_role.eks_service_role.name
  policy_arn = "arn:aws:iam::aws:policy/AmazonEKSClusterPolicy"
}

resource "aws_iam_role_policy_attachment" eks_service_role_AmazonEKSServicePolicy {
  role = aws_iam_role.eks_service_role.name
  policy_arn = "arn:aws:iam::aws:policy/AmazonEKSServicePolicy"
}

resource "aws_eks_cluster" "test-cluster" {
  name     = "test-cluster"
  role_arn = aws_iam_role.eks_service_role.arn

  vpc_config {
    subnet_ids = concat(aws_subnet.private[*].id)
  }
}
```

It appears that the IAM Role permissions (the two `aws_iam_role_policy_attachment` resources) are not explicitly declared as dependencies to be created before and deleted after the EKS Cluster itself via [`depends_on`](https://www.terraform.io/docs/configuration/resources.html#depends_on-explicit-resource-dependencies). Its likely that the removal of the IAM Role permissions concurrently with the the start of the EKS Cluster removal prevents the EKS service from managing its underlying infrastructure. The below configuration change for the `aws_eks_cluster` resource should resolve the issue in this case:

```hcl
resource "aws_eks_cluster" "test-cluster" {
  name     = "test-cluster"
  role_arn = aws_iam_role.eks_service_role.arn

  vpc_config {
    subnet_ids = concat(aws_subnet.private[*].id)
  }

  # Ensure that IAM Role permissions are created before and deleted after EKS Cluster handling.
  # Otherwise, EKS will not be able to properly delete EC2 infrastructure such as managed Security Groups
  depends_on = [
    aws_iam_role_policy_attachment.eks_service_role_AmazonEKSClusterPolicy,
    aws_iam_role_policy_attachment.eks_service_role_AmazonEKSServicePolicy,
  ]
}
```

I'm going to leave this issue open to note that we should update the `aws_eks_cluster` resource documentation to mention the above (similar to https://github.com/terraform-providers/terraform-provider-aws/pull/11010).

---

In the future, we can help reduce the need for explicit Terraform dependencies such as these via supporting the management of attached IAM Role policies directly in the `aws_iam_role` resource (e.g. #5904).

**Comment by hangxie** [comment]
Confirmed it fixed the problem by specifying the right dependency.

Thank you, @bflad

**Comment by bflad** [comment]
The documentation update to include extra information about the `depends_on` usage has been merged and will publish when version 2.41.0 of the Terraform AWS Provider is released this week. 👍
